### ISSUE TITLE
Possible race condition in WriterFactory handler cache

### ISSUE THREAD

**Issue by favila** [cognitect/transit-java#20]
Clojurescript experienced an exception when many file writes of transit data happen simultaneously, see [CLJS-1759](http://dev.clojure.org/jira/browse/CLJS-1759). The problem was worked around by serializing all use of transit. (i.e. only write one file at a time.)

A followup ticket [CLJS-1761](http://dev.clojure.org/jira/browse/CLJS-1761) was created to investigate the reason the workaround was necessary.

In the stack trace I noticed that the exception rethrown out of the `write()` is `java.lang.Exception: Not supported: class java.lang.Integer`, which sounds like a missing handler, but for a type which has a default handler. This led me to the hunch that [these lines](https://github.com/cognitect/transit-java/blob/3f359c434264f675460bc2662dde2a1b2d8e9559/src/main/java/com/cognitect/transit/impl/WriterFactory.java#L74-L76) are a race condition. `containsKey()` should not be used, just `get()` and a null check. I'm also not sure the following synchronization shouldn't be on the cache itself instead of the WriterFactory.

Both the `handlerMap()` and `verboseHandlerMap()` methods appear to have the same races.

However, I still cannot explain why, in ClojureScript's case, the writeHandlerCache would ever have more than one entry because it is repeatedly invoked with the same (identical) custom map handler object. Since it never has more than one entry, it should never evict anything, so this race condition would never be triggered.

**Comment by favila** [comment]
ReaderFactory's cache appears to have [a similar race](https://github.com/cognitect/transit-java/blob/3f359c434264f675460bc2662dde2a1b2d8e9559/src/main/java/com/cognitect/transit/impl/ReaderFactory.java#L62-L64). Perhaps all uses of the Cache class need auditing.

**Comment by mfikes** [comment]
I found a change in `transit-java` that fixes the issue for me (the change is right at the bottom of this comment). 

I also attempted to create a minimal repro, separate from the project I am experiencing it in, but failed to achieve that. So, I'll attempt to be careful below in explaining the facts that I've observed _vs._ my theory / speculation that let me to try the change.

As @favila indicates, I'm seeing this with CLJS-1759. A high level summary of this is that is involves using the ClojureScript compiler with its `:parallel-build` compiler optimization enabled, with the newer variant of the ClojureScript compiler that writes out its analysis cache data using Transit (as opposed to EDN).

The project I'm experiencing it in is in the ClojureScript build portion of  [Planck](https://github.com/mfikes/planck). I'm seeing it on a dual-hexacore OS X box, and it appears to occur when 3 threads are simultaneously attempting to write. It doesn't happen all of the time, but it does occur on a good fraction of the builds, and appears to affect two namespaces in particular. Here are transcripts showing it happening: https://gist.github.com/mfikes/e3313ec059b604674f60d326e701a7f1

Perhaps of interest is that the ClojureScript compiler is employing a fixed thread pool and is also passing in a custom write handler map.

Sketchy theory: @swannodette and I, at a high level, fundamentally don't expect any issues as new writers are created in each thread, thus resulting in avoiding any data structure sharing. But I think I see some sharing in `transit-java`: ClojureScript is always passing in the exact same custom handlers map (`cljs.analyzer/transit-write-opts`), which is used as a key in a cache map. This results in a `WriterHandlerMap` being constructed and used by all threads and they all share the result of calling `WriterFactory.defaultHandlers()`. 

I added a bit of code 

```
   System.err.println(writeHandlerMap.keySet());
```

to print out the keys in the handlers map when things fail (which is on this line: https://github.com/cognitect/transit-java/blob/3f359c434264f675460bc2662dde2a1b2d8e9559/src/main/java/com/cognitect/transit/impl/AbstractEmitter.java#L176) and I saw this in one of the stack traces that resulted when it was trying to print the keys of the write handlers map:

```
Caused by: java.util.ConcurrentModificationException
       	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1437)
       	at java.util.HashMap$KeyIterator.next(HashMap.java:1461)
       	at java.util.AbstractCollection.toString(AbstractCollection.java:461)
       	at java.lang.String.valueOf(String.java:2994)
       	at java.io.PrintStream.println(PrintStream.java:821)
       	at com.cognitect.transit.impl.AbstractEmitter.marshal(AbstractEmitter.java:176)
```

So, the theory is that something else is modifying this map, and it is not thread safe (it is based on a `HashMap`). 

Another interesting theory is that, effectively the `HashMap` created by `WriterFactory.defaultHandlers()` is not being properly published from one thread to another (in JCiP parlance). You can see that this might be the case because the `Cache` is just an LRU with no synchronization primitives, and a check and then-lock pattern is used. In short this could cause the `HashMap` to be partially seen by another thread, thus leading to the "initial nulls , initial zeros" being seen. (This second theory is part of what led me to make the speculative change.)

Back to non-theory, back to facts: Without changing anything else, I changed the last line of `WriterFactory.defaultHandlers()` to be:

```
return Collections.synchronizedMap(handlers);
```

and the problems disappeared.

**Comment by dchelimsky** [comment]
Hi @favila and @mfikes. Thanks to both of you for all of the careful analysis of the problem.

@favila I don't think the issue is caused by using `containsKey` there because if it returns true then the value is there and `get()` works as expected, and if it's false then we fall into the synchronized block. If another thread was in the middle of modifying the handler cache, the current thread will wait for that synchronized modification to complete, after which the synchronized call to `containsKey` will return true. The idea of having the unsynchronized `containsKey`/`get` was to avoid the synchronization hit for all of the calls to `handlerMap` after the first call with any given value. That all make sense?

I can see some logic in synchronizing on the handler cache, but I don't understand how that would actually fix anything here either.

I think it's possible that the issue is publishing a partially constructed object, as @mfikes suggests, though I'm unclear as to whether culprit is the map itself or one of the objects in the map, since all the values are `new xxx`. Of course, all of those handlers are actually immutable themselves, so maybe they could be singletons. Need to think about that.

In any case, I'm going to talk this over w/ a couple of other folks and follow up here later this week.

**Comment by mfikes** [comment]
Thanks @dchelimsky. Since I've been unable to produce a minimal repro, if you ultimately end up with an idea you'd like to experiment with (perhaps something that avoids the heavy-handed `synchronizedMap`), I'd be more than happy to either give code a spin in my environment, or even share precise details on how to reproduce the issue, if that were desired—it is possible to reproduce using publicly-available code in the Planck repository.

**Comment by mfikes** [comment]
Here is how I reproduce the issue:

```
git clone https://github.com/mfikes/planck
cd planck
git reset --hard f3a426596247653d76293427fd89b0471ed65f4e
cd planck-cljs
while true; do script/clean; script/build; done
```

I can reproduce the issue within a few build iterations (in other words, within few minutes minutes) on this box:

> Mac Pro (Mid 2012), 2 x 2.4 GHz 6-Core Intel Xeon

but I absolutely cannot reproduce it on this box:

> MacBook Pro (Retina, 15-inch, Mid 2015), 2.5 GHz Intel Core i7

so my guess is that you need to be on a box with a lot of physical cores to see it.

**Comment by favila** [comment]
This is an attempt at a [minimal repro](https://gist.github.com/favila/fbd3348a6666386ee063bf1981ad5541). It just tries to stress the WriteFactory handler caches on as many threads as possible. Running it on an old 4-core i3, oracle java 1.8.0-45 does not cause problems, however, and I do not have a bigger machine available right now. Perhaps you @mfikes could give it a try?

**Comment by dchelimsky** [comment]
Thanks for the repros. I probably won't be able to devote time to this until Friday, but I'll keep you updated.

**Comment by mfikes** [comment]
@favila I tried with the minimal repro and failed. I edited the file to eliminate its package and the extra `run` and here is what it did. It spun the CPUs but didn't print anything:
```
$ java -version
java version "1.8.0_102"
Java(TM) SE Runtime Environment (build 1.8.0_102-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.102-b14, mixed mode)
$ javac -cp /Users/mfikes/.m2/repository/com/cognitect/transit-java/0.8.313/transit-java-0.8.313.jar ParallelStressTest.java
Note: ParallelStressTest.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
$ java -classpath /Users/mfikes/.m2/repository/com/cognitect/transit-java/0.8.313/transit-java-0.8.313.jar:/Users/mfikes/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.3.2/jackson-core-2.3.2.jar:. ParallelStressTest
```

**Comment by dchelimsky** [comment]
Hi @mfikes and @favila. Here is a possible culprit: https://github.com/cognitect/transit-java/blob/1ee9d2100cc9770756f6be2dca600d807bed5a00/src/main/java/com/cognitect/transit/impl/WriteHandlerMap.java#L141-L150. That method actually modifies the handlers map as it is writing when it encounters a type that is not yet registered, but is an extension of a type that is. 

This _could_ explain the ConcurrentModificationException @mfikes reported earlier this thread, in which case `synchronizedMap` _might_ solve the problem. Out of curiosity, does your app try to serialize subtypes of types supported by transit?

**Comment by mfikes** [comment]
@dchelimsky This is the ClojureScript compiler using transit (not really my app proper). I added two places to print the registering of handlers, and it logged registrations, but perhaps these are normal?

```
$ script/build
Leiningen 2.6.1 on Java 1.8.0_102 Java HotSpot(TM) 64-Bit Server VM
Building
checkBaseInterfaces: Registering a handler for class clojure.lang.PersistentHashMap
checkBaseInterfaces: Registering a handler for class clojure.lang.PersistentHashMap
checkBaseInterfaces: Registering a handler for class clojure.lang.PersistentHashMap
checkBaseInterfaces: Registering a handler for class clojure.lang.PersistentArrayMap
checkBaseInterfaces: Registering a handler for class clojure.lang.PersistentArrayMap
checkBaseInterfaces: Registering a handler for class clojure.lang.PersistentArrayMap
checkBaseInterfaces: Registering a handler for class clojure.lang.PersistentHashSet
checkBaseInterfaces: Registering a handler for class com.cognitect.transit.impl.TaggedValueImpl
checkBaseInterfaces: Registering a handler for class clojure.lang.PersistentList
checkBaseInterfaces: Registering a handler for class clojure.lang.PersistentVector
checkBaseInterfaces: Registering a handler for class clojure.lang.PersistentVector$ChunkedSeq
checkBaseInterfaces: Registering a handler for class clojure.lang.LazySeq
WARNING: boolean? already refers to: cljs.core/boolean? being replaced by: fipp.visit/boolean? at line 32 out/fipp/visit.cljc
Exception in thread "main" clojure.lang.ExceptionInfo: failed compiling file:out/clojure/core/rrb_vector/protocols.cljs {:file #object[java.io.File 0x30ca66c5 "out/clojure/core/rrb_vector/protocols.cljs"]}, compiling:(/Users/mfikes/Desktop/planck/planck-cljs/script/build.clj:16:1)
       	at clojure.lang.Compiler.load(Compiler.java:7391)
       	at clojure.lang.Compiler.loadFile(Compiler.java:7317)
       	at clojure.main$load_script.invokeStatic(main.clj:275)
       	at clojure.main$script_opt.invokeStatic(main.clj:335)
       	at clojure.main$script_opt.invoke(main.clj:330)
       	at clojure.main$main.invokeStatic(main.clj:421)
       	at clojure.main$main.doInvoke(main.clj:384)
       	at clojure.lang.RestFn.invoke(RestFn.java:408)
       	at clojure.lang.Var.invoke(Var.java:379)
       	at clojure.lang.AFn.applyToHelper(AFn.java:154)
       	at clojure.lang.Var.applyTo(Var.java:700)
       	at clojure.main.main(main.java:37)
Caused by: clojure.lang.ExceptionInfo: failed compiling file:out/clojure/core/rrb_vector/protocols.cljs {:file #object[java.io.File 0x30ca66c5 "out/clojure/core/rrb_vector/protocols.cljs"]}
       	at clojure.core$ex_info.invokeStatic(core.clj:4617)
       	at clojure.core$ex_info.invoke(core.clj:4617)
       	at cljs.compiler$compile_file$fn__3456.invoke(compiler.cljc:1386)
       	at cljs.compiler$compile_file.invokeStatic(compiler.cljc:1352)
       	at cljs.compiler$compile_file.invoke(compiler.cljc:1332)
       	at cljs.closure$compile_file.invokeStatic(closure.clj:474)
       	at cljs.closure$compile_file.invoke(closure.clj:465)
       	at cljs.closure$eval5203$fn__5204.invoke(closure.clj:541)
       	at cljs.closure$eval5139$fn__5140$G__5128__5147.invoke(closure.clj:431)
       	at cljs.closure$compile_from_jar.invokeStatic(closure.clj:523)
       	at cljs.closure$compile_from_jar.invoke(closure.clj:511)
       	at cljs.closure$eval5209$fn__5210.invoke(closure.clj:551)
       	at cljs.closure$eval5139$fn__5140$G__5128__5147.invoke(closure.clj:431)
       	at cljs.closure$compile_task$fn__5294.invoke(closure.clj:821)
       	at cljs.closure$compile_task.invokeStatic(closure.clj:819)
       	at cljs.closure$compile_task.invoke(closure.clj:812)
       	at cljs.closure$parallel_compile_sources$fn__5300.invoke(closure.clj:848)
       	at clojure.lang.AFn.applyToHelper(AFn.java:152)
       	at clojure.lang.AFn.applyTo(AFn.java:144)
       	at clojure.core$apply.invokeStatic(core.clj:646)
       	at clojure.core$with_bindings_STAR_.invokeStatic(core.clj:1881)
       	at clojure.core$with_bindings_STAR_.doInvoke(core.clj:1881)
       	at clojure.lang.RestFn.invoke(RestFn.java:425)
       	at clojure.lang.AFn.applyToHelper(AFn.java:156)
       	at clojure.lang.RestFn.applyTo(RestFn.java:132)
       	at clojure.core$apply.invokeStatic(core.clj:650)
       	at clojure.core$bound_fn_STAR_$fn__4671.doInvoke(core.clj:1911)
       	at clojure.lang.RestFn.invoke(RestFn.java:397)
       	at clojure.lang.AFn.run(AFn.java:22)
       	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
       	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
       	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: java.lang.Exception: Not supported: class java.lang.Integer
       	at com.cognitect.transit.impl.WriterFactory$1.write(WriterFactory.java:129)
       	at cognitect.transit$write.invokeStatic(transit.clj:149)
       	at cognitect.transit$write.invoke(transit.clj:146)
       	at cljs.analyzer$write_analysis_cache.invokeStatic(analyzer.cljc:3127)
       	at cljs.analyzer$write_analysis_cache.invoke(analyzer.cljc:3114)
       	at cljs.compiler$emit_source.invokeStatic(compiler.cljc:1283)
       	at cljs.compiler$emit_source.invoke(compiler.cljc:1232)
       	at cljs.compiler$compile_file_STAR_$fn__3433.invoke(compiler.cljc:1304)
       	at cljs.compiler$with_core_cljs.invokeStatic(compiler.cljc:1154)
       	at cljs.compiler$with_core_cljs.invoke(compiler.cljc:1145)
       	at cljs.compiler$compile_file_STAR_.invokeStatic(compiler.cljc:1293)
       	at cljs.compiler$compile_file_STAR_.invoke(compiler.cljc:1289)
       	at cljs.compiler$compile_file$fn__3456.invoke(compiler.cljc:1374)
       	... 29 more
Caused by: java.lang.Exception: Not supported: class java.lang.Integer
       	at com.cognitect.transit.impl.AbstractEmitter.marshal(AbstractEmitter.java:176)
       	at com.cognitect.transit.impl.JsonEmitter.emitMap(JsonEmitter.java:158)
       	at com.cognitect.transit.impl.AbstractEmitter.emitMap(AbstractEmitter.java:70)
       	at com.cognitect.transit.impl.AbstractEmitter.marshal(AbstractEmitter.java:166)
       	at com.cognitect.transit.impl.JsonEmitter.emitMap(JsonEmitter.java:158)
       	at com.cognitect.transit.impl.AbstractEmitter.emitMap(AbstractEmitter.java:70)
       	at com.cognitect.transit.impl.AbstractEmitter.marshal(AbstractEmitter.java:166)
       	at com.cognitect.transit.impl.JsonEmitter.emitMap(JsonEmitter.java:158)
       	at com.cognitect.transit.impl.AbstractEmitter.emitMap(AbstractEmitter.java:70)
       	at com.cognitect.transit.impl.AbstractEmitter.marshal(AbstractEmitter.java:166)
       	at com.cognitect.transit.impl.AbstractEmitter.marshalTop(AbstractEmitter.java:193)
       	at com.cognitect.transit.impl.JsonEmitter.emit(JsonEmitter.java:28)
       	at com.cognitect.transit.impl.WriterFactory$1.write(WriterFactory.java:126)
       	... 41 more
```

**Comment by dchelimsky** [comment]
Hi again @mfikes and @favila. The internal handler maps get updated whenever the writer encounters an object with an unregistered type that is a subtype of a registered type, which could be the source of the issues you're seeing. One solution is to store unmodifiable maps in the cache, and then each instance of a WriteHanderMap would get its own, separate map for these additional mappings, all hidden behind `getHandler(o)` on the WriteHandlerMap. This has some implications across other parts of the API and other APIs (i.e. transit-clj) and time is thin right now so it might be a week or two before we can actually show any progress, but we're working on it and will follow up as we have new information.

**Comment by mfikes** [comment]
Thanks @dchelimsky. 

FWIW, for the time being, the latest releases of the ClojureScript compiler include a mutex so that Transit writes never occur concurrently, effectively working around the issue.

**Comment by dchelimsky** [comment]
@mfikes I posted a PR that I believe addresses this issue, though I can't prove it because I've not been able to repro it :) Would you please grab that branch and test it? Comments welcome as well, of course! Cheers! /cc @favila

**Comment by mfikes** [comment]
@dchelimsky absolutely ... it will take me perhaps a few days to get to it (on travel)

**Comment by mfikes** [comment]
@dchelimsky When I attempt to repro with a locally build copy of the branch in #21 things work just fine now. Thanks for working on it!

**Comment by dchelimsky** [comment]
Fixed in transit-java-0.8.316 and transit-clj-0.8.290

**Comment by mfikes** [comment]
Confirmed fixed downstream using transit-clj-0.8.290
